lib/msal-node/src/cache/CacheManager.ts

```diff
@@ -0,0 +1,229 @@
+                JSON.parse(this.cacheSnapshot),
+                finalState
+            );
+        }
+        this.hasChanged = false;
+
+        return JSON.stringify(finalState);
+    }
+
+    /**
+     * Deserializes JSON to in-memory cache. JSON should be in MSAL cache schema format
+     * @param cache
+     */
+    deserialize(cache: string): void {
+        this.cacheSnapshot = cache;
+
+        if (!StringUtils.isEmpty(this.cacheSnapshot)) {
+            const deserializedCache = Deserializer.deserializeAllCache(this.overlayDefaults(JSON.parse(this.cacheSnapshot)));
+            this.storage.setCache(deserializedCache);
+        }
+    }
+
+    /**
+     * Serializes cache into JSON and calls ICachePlugin.writeToStorage. ICachePlugin must be set on ClientApplication
+     */
+    async writeToPersistence(): Promise<void> {
+        if (this.persistence) {
+            let cache = Serializer.serializeAllCache(this.storage.getCache() as InMemoryCache);
+
+            const getMergedState = (stateFromDisk: string) => {
+
+                if (!StringUtils.isEmpty(stateFromDisk)) {
+                    this.cacheSnapshot = stateFromDisk;
+                    cache = this.mergeState(JSON.parse(stateFromDisk), cache);
+                }
+
+                return JSON.stringify(cache);
+            };
+
+            await this.persistence.writeToStorage(getMergedState);
+            this.hasChanged = false;
+        } else {
+            throw ClientAuthError.createCachePluginError();
+        }
+    }
+
+    /**
+     * Calls ICachePlugin.readFromStorage and deserializes JSON to in-memory cache.
+     * ICachePlugin must be set on ClientApplication.
+     */
+    async readFromPersistence(): Promise<void> {
+        if (this.persistence) {
+            this.cacheSnapshot = await this.persistence.readFromStorage();
+
+            if (!StringUtils.isEmpty(this.cacheSnapshot)) {
+                const cache = this.overlayDefaults(JSON.parse(this.cacheSnapshot));
+                const deserializedCache = Deserializer.deserializeAllCache(cache);
+                this.storage.setCache(deserializedCache);
+            }
+        } else {
+            throw ClientAuthError.createCachePluginError();
+        }
+    }
+
+    /**
+     * Called when the cache has changed state.
+     */
+    private handleChangeEvent() {
+        this.hasChanged = true;
+    }
+
+    /**
+     * Merge in memory cache with the cache snapshot.
+     * @param oldState
+     * @param currentState
+     */
+    private mergeState(oldState: JsonCache, currentState: JsonCache): JsonCache {
+        let stateAfterRemoval = this.mergeRemovals(oldState, currentState);
+        return this.mergeUpdates(stateAfterRemoval, currentState);
+    }
+
+    /**
+     * Deep update of oldState based on newState values
+     * @param oldState
+     * @param newState
+     */
+    private mergeUpdates(oldState: any, newState: any): JsonCache {
+
+        let finalState = oldState;
```

Comment:
Can you create a copy instead? `let finalState = { ...oldState };`

Reply:
done

Reply:
Was this change pushed? Not seeing it.